torch/distributed/rpc/api.py

```diff
@@ -333,9 +335,17 @@ def shutdown(graceful=True, timeout=DEFAULT_SHUTDOWN_TIMEOUT):
     """
     if graceful:
         try:
-            _wait_all_workers(timeout)
-            _delete_all_user_and_unforked_owner_rrefs()
-            _get_current_rpc_agent().join(shutdown=True)
+            agent = _get_current_rpc_agent()
+            if agent.is_static_group:
+                _wait_all_workers(timeout)
+                _delete_all_user_and_unforked_owner_rrefs()
+                agent.join(shutdown=True)
+            else:
+                # This is a dynamic group so we need to grab the token for the operation
+                workerInfo = agent.get_worker_info()
+                with utils.group_membership_management(agent._get_store(), workerInfo.name):
+                    agent.join(shutdown=True)
```

Comment:
I assume this part is WIP, as a global join won't work here?

Reply:
Yes! The join is not global anymore. Now if TPagent is part of a dynamic group, then it will leave the group when there are no active calls remaining for it.